django/db/models/fields/json.py

```diff
@@ -0,0 +1,528 @@
+import json
+
+from django import forms
+from django.conf import settings
+from django.core import checks, exceptions
+from django.db import connections
+from django.db.models import lookups
+from django.db.models.lookups import (
+    FieldGetDbPrepValueMixin, Lookup, Transform,
+)
+from django.db.utils import NotSupportedError
+from django.utils.translation import gettext_lazy as _
+
+from . import Field
+from .mixins import CheckFieldDefaultMixin
+
+
+class JSONField(CheckFieldDefaultMixin, Field):
+    description = _('A JSON object')
+    default_error_messages = {
+        'invalid': _('Value must be valid JSON.'),
+    }
+    _default_hint = ('dict', '{}')
+
+    def __init__(
+        self, verbose_name=None, name=None, encoder=None,
+        decoder=None, *args, **kwargs
+    ):
+        self.encoder, self.decoder = encoder, decoder
+        super().__init__(verbose_name, name, *args, **kwargs)
+
+    def check(self, **kwargs):
+        return [
+            *super().check(**kwargs),
+            *self._check_json_support(),
+        ]
+
+    @classmethod
+    def _check_json_support(cls):
+        errors = []
+        any_jsonfield_support = False
+        for db in settings.DATABASES:
+            connection = connections[db]
+            if connection.features.supports_json_field:
+                any_jsonfield_support = True
+        if not any_jsonfield_support:
+            errors.append(
+                checks.Error(
+                    'No database connection that supports JSONField is found.',
+                    hint='See the documentation of JSONField for supported database versions.',
+                    obj=cls
+                )
+            )
+        return errors
+
+    def deconstruct(self):
+        name, path, args, kwargs = super().deconstruct()
+        path = 'django.db.models.JSONField'
+        if self.encoder is not None:
+            kwargs['encoder'] = self.encoder
+        if self.decoder is not None:
+            kwargs['decoder'] = self.decoder
+        return name, path, args, kwargs
+
+    def from_db_value(self, value, expression, connection):
+        if value is None:
+            return value
+        elif connection.features.interprets_empty_strings_as_nulls and value == '':
+            return None
```

Comment:
These two lines should be removed, because they don't reflect Django's behavior in this case, please check [null-and-empty-strings](https://docs.djangoproject.com/en/2.2/ref/databases/#null-and-empty-strings).